Update the version generation logic and info

tuhaihe opened the pull request:
<!-- Thank you for your contribution to Apache Cloudberry (Incubating)! -->

Fow now, if we build Cloudberry from a source zip file (not using git
clone), the version information displayed is not matched the right info,
always return 1.0.0 like this:

```
   PostgreSQL 14.4 (Apache Cloudberry 1.0.0 build dev)
```

In this PR, these changes are made:

- Update the PACKAGE_VERSION string from 1.0.0 to to 2.0.0-devel
- Generate the new configure file with the new configure.ac
- Update the getversion file to take the PACKAGE_VERSION as the base
  version info, and append `+dev.<count>.g<sha>` whenever possible
  unless the current commit is exactly on a tag.

After these changes, will return the right info no matter how they get
the source code, like git-clone, source tarballs.

For each release, we also need to update the version string to match the
right release version manually in each branch.

See: http://github.com/apache/cloudberry/discussions/909

### What does this PR do?
<!-- Brief overview of the changes, including any major features or fixes -->

### Type of Change
- [ ] Bug fix (non-breaking change)
- [ ] New feature (non-breaking change)
- [ ] Breaking change (fix or feature with breaking changes)
- [ ] Documentation update

### Breaking Changes
<!-- Remove if not applicable. If yes, explain impact and migration path -->

### Test Plan
<!-- How did you test these changes? -->
- [ ] Unit tests added/updated
- [ ] Integration tests added/updated
- [ ] Passed `make installcheck`
- [ ] Passed `make -C src/test installcheck-cbdb-parallel`

### Impact
<!-- Remove sections that don't apply -->
**Performance:**
<!-- Any performance implications? -->

**User-facing changes:**
<!-- Any changes visible to users? -->

**Dependencies:**
<!-- New dependencies or version changes? -->

### Checklist
- [ ] Followed [contribution guide](https://cloudberry.apache.org/contribute/code)
- [ ] Added/updated documentation
- [ ] Reviewed code for security implications
- [ ] Requested review from [cloudberry committers](https://github.com/orgs/apache/teams/cloudberry-committers)

### Additional Context
<!-- Any other information that would help reviewers? Remove if none -->

### CI Skip Instructions
<!--
To skip CI builds, add the appropriate CI skip identifier to your PR title.
The identifier must:
- Be in square brackets []
- Include the word "ci" and either "skip" or "no"
- Only use for documentation-only changes or when absolutely necessary
-->

---
<!-- Join our community:
- Mailing list: [dev@cloudberry.apache.org](https://lists.apache.org/list.html?dev@cloudberry.apache.org) (subscribe: dev-subscribe@cloudberry.apache.org)
- Discussions: https://github.com/apache/cloudberry/discussions -->

tuhaihe commented:
> Cloudberry Database need a version, but not main .

Agreed. For PG, I found they always use the next `release version + devel`, like `18devel`. It's hard for us to define the next major version, right?  One simple way is to set the main branch release as 99.0.0 or 999.0.0?

https://github.com/postgres/postgres/blob/7da344b9f84f0c63590a34136f3fa5d0ab128657/configure.ac#L20

tuhaihe commented:
This PR is ready for review now. This PR has two big changes:

- Add the logic to choose the newer version as the final version
- Adopt the Postgres way, letting the main branch be in the latest version. For each release, need to update the version manually.
  * https://github.com/postgres/postgres/commit/e26810d01d441a457217a6eae9c2989fba29b80f
  * https://github.com/postgres/postgres/commit/6304632eaa2107bb1763d29e213ff166ff6104c0

tuhaihe commented:
If users build the cloudberry using `git clone --depth 1 https://github.com/apache/cloudberry.git`, they also will not receive the correct version information prior to this PR.

edespino reviewed (changes requested):
The version logic added in this commit introduces an issue for builds between annotated tags.

Specifically, when `PACKAGE_VERSION` (e.g. `2.0.0devel`) is newer than the last Git tag, the logic ignores the `git describe` output — and we lose valuable metadata like the number of commits since the tag and the commit SHA.

This means:

- We get a version like `2.0.0devel`, even if we’re 52 commits past the last tag
- It's harder to trace builds to a specific commit
-  CI builds between tags are indistinguishable from tagged releases

While comparing versions to prefer the latest is understandable, we believe it’s more useful to always include commit SHA and commit distance when the current HEAD is not on a tag.

Suggested fix: keep `PACKAGE_VERSION` as the base, but append `+dev.<count>.g<sha>` whenever possible unless the current commit is exactly on a tag.

This hybrid approach:

- Keeps version strings clean for releases
- Adds valuable traceability during development
- Matches behavior seen in projects like PostgreSQL and Rust

Happy to submit a patch if helpful!

tuhaihe commented:
Hey, @edespino thanks for your valuable feedback. I made updates to my PR after considering your suggestions. Please take a look again.

I also did some tests under different environments. The following are the outputs:

1. Testing via `git clone` the source code with full commit log: 

```shell
[gpadmin@cdw cloudberry]$ /usr/local/cloudberry-db/bin/postgres --gp-version
postgres (Apache Cloudberry) 2.0.0devel+dev.1533.gc73181c522 build dev

[gpadmin@cdw cloudberry]$ ./getversion
2.0.0devel+dev.1533.gc73181c522 build dev

[gpadmin@cdw cloudberry]$ psql template1 -c 'SELECT version()'
                                                                                                            version

----------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------
 PostgreSQL 14.4 (Apache Cloudberry 2.0.0devel+dev.1533.gc73181c522 build dev) on x86_64-pc-linux-gnu, compiled by gcc (GCC) 11.5.0 20240719 (
Red Hat 11.5.0-5), 64-bit compiled on Apr  6 2025 23:09:03 (with assert checking)
(1 row)
```

2. Testing via downloading the source .tar file:
```
[gpadmin@cdw cloudberry]$ /usr/local/cloudberry-db/bin/postgres --gp-version
postgres (Apache Cloudberry) 2.0.0devel build dev

[gpadmin@cdw cloudberry]$ ./getversion
2.0.0devel build dev

[gpadmin@cdw cloudberry]$ psql template1 -c 'SELECT version()'
                                                                                                  version

----------------------------------------------------------------------------------------------------------------------------------------------
-------------------------------------------------------------
 PostgreSQL 14.4 (Apache Cloudberry 2.0.0devel build dev) on x86_64-pc-linux-gnu, compiled by gcc (GCC) 11.5.0 20240719 (Red Hat 11.5.0-5), 64
-bit compiled on Apr  6 2025 23:22:50 (with assert checking)
(1 row)
```

3. Testing when the current HEAD is on a tag
```
[gpadmin@cdw cloudberry]$ ./getversion
2.0.0devel+dev.1533.gc73181c522 build dev

[gpadmin@cdw cloudberry]$ git config --global user.email "test@cloudberry.a.org"
[gpadmin@cdw cloudberry]$ git config --global user.name "Dianjin Wang"
[gpadmin@cdw cloudberry]$ git tag -a 2.0.0 c73181c522056878361e3279c3934043fd27a4d9 -m "Test 2.0.0 tag"
[gpadmin@cdw cloudberry]$ ./getversion
2.0.0devel build dev

[gpadmin@cdw cloudberry]$ /usr/local/cloudberry-db/bin/postgres --gp-version
postgres (Apache Cloudberry) 2.0.0devel build dev

[gpadmin@cdw cloudberry]$ psql template1 -c 'SELECT version()'
                                                                                                  version

----------------------------------------------------------------------------------------------------------------------------------------------
-------------------------------------------------------------
 PostgreSQL 14.4 (Apache Cloudberry 2.0.0devel build dev) on x86_64-pc-linux-gnu, compiled by gcc (GCC) 11.5.0 20240719 (Red Hat 11.5.0-5), 64
-bit compiled on Apr  6 2025 23:35:50 (with assert checking)
(1 row)
```

avamingli commented on configure.ac:
```diff
@@ -31,7 +31,7 @@ dnl your responsibility whether the result works or not.])])
 AC_CONFIG_SRCDIR([src/backend/access/common/heaptuple.c])
 AC_CONFIG_AUX_DIR(config)
-AC_PREFIX_DEFAULT(/usr/local/cbdb)
+AC_PREFIX_DEFAULT(/usr/local/cloudberry)
```
This is a default path, not a word. Change it will make unnecessary troubles, please recover that.

tuhaihe replied:
All right. has been updated in the latest commit https://github.com/apache/cloudberry/pull/929/commits/2e0efb7fce5f00ada10d0327c8ecd41e9462f3f7

tuhaihe commented:
Hi guys, now this PR is ready for review again. Thanks for your help! @edespino @yjhjstz @avamingli

tuhaihe commented:
Hey @jiaqizho I use the command `autoreconf -i` to regenerate the configure file after updating the configure.ac. Some PAX-related options have changed. Could you help take a look to see if the changes are okay?

The changes are in the latest commit:https://github.com/apache/cloudberry/pull/929/commits/25e4e792121e729852e25733ce73e6e2cd761223

jiaqizho commented on getversion:
```diff
@@ -23,10 +23,11 @@ if type git >/dev/null 2>&1 && [ -d '.git' ] ; then
-		VERSION+=+
-		VERSION+=$(git rev-parse --short HEAD)
+		GIT_VERSION=$(generate_dev_version)
+	# Append commit distance and SHA if not on a tag
```
align this line

tuhaihe replied:
Done. Thanks for your comment.

edespino commented on getversion:
```diff
@@ -23,10 +23,11 @@ if type git >/dev/null 2>&1 && [ -d '.git' ] ; then
-	else
-		VERSION+=+
-		VERSION+=$(git rev-parse --short HEAD)
+		GIT_VERSION=$(generate_dev_version)
+	    # Append commit distance and SHA if not on a tag
+		if ! git describe --exact-match >/dev/null 2>&1; then
+			VERSION+=\+$(echo "$GIT_VERSION" | cut -d'+' -f2-)
+		fi
```
If HEAD is on a tag the tag value should be used as the version. Even in your example, "3. Testing when the current HEAD is on a tag", you can see the tag is not used.

tuhaihe replied:
@edespino PTAL again, thanks!

edespino reviewed (changes requested):
This still needs work. When HEAD is on a tag, the tag is not used for the version.

tuhaihe commented:
New testing when the current HEAD is on a tag:
```

[gpadmin@cdw cloudberry]$ ./getversion
1.6.0+dev.1932.g5d0e2ef730 build dev
[gpadmin@cdw cloudberry]$ git switch tuhaihe/match-release-version
Switched to branch 'tuhaihe/match-release-version'
[gpadmin@cdw cloudberry]$ ./getversion
2.0.0devel+dev.1933.g9e4166e0d6 build dev
[gpadmin@cdw cloudberry]$ git config --global user.email "test@cloudberry.a.org"
[gpadmin@cdw cloudberry]$ git config --global user.name "Dianjin Wang"
[gpadmin@cdw cloudberry]$ git tag -a 2.0.0 9e4166e0d6 -m "Test 2.0.0 tag"
[gpadmin@cdw cloudberry]$ ./getversion
2.0.0 build dev



[gpadmin@cdw cloudberry]$ /usr/local/cloudberry-db/bin/postgres --gp-version
postgres (Apache Cloudberry) 2.0.0 build dev

[gpadmin@cdw cloudberry]$ psql template1 -c 'SELECT version()'
                                                                                               version

-----------------------------------------------------------------------------------------------------------------------------
-------------------------------------------------------------------------
 PostgreSQL 14.4 (Apache Cloudberry 2.0.0 build dev) on x86_64-pc-linux-gnu, compiled by gcc (GCC) 11.5.0 20240719 (Red Hat 1
1.5.0-5), 64-bit compiled on Apr 29 2025 21:12:49 (with assert checking)
(1 row)
```

tuhaihe commented:
I droped this commit https://github.com/apache/cloudberry/commit/25e4e792121e729852e25733ce73e6e2cd761223 as we will disable the PAX option by default.

edespino commented:
@tuhaihe  Here is my version of getversion. There was an issue if previous annotated tag had a "+" character.

example:
     git tag -a 2.0.0+RC1 -m "Apache Cloudberry (Incubating) 2.0.0 RC1"

This version of the script handles that case. Please review when you get a chance.

```
#!/usr/bin/env bash
# ======================================================================
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# -----------------------------------------------------------------------------
# getversion — determine the project version for packaging and diagnostics
# -----------------------------------------------------------------------------
#
# This script computes a version string using the following logic:
#
#   1. Extract the base PACKAGE_VERSION from the `configure` script.
#
#   2. If running inside a Git repository and `git describe` is usable:
#
#      a. If the current commit matches an annotated tag exactly:
#           VERSION = <tag>
#         e.g.,     VERSION = 2.0.0+RC1
#
#      b. If the current commit is ahead of a tag:
#           VERSION = <PACKAGE_VERSION>+dev.<N>.g<commit>
#         e.g.,     VERSION = 2.0.1devel+dev.3.gabcd123
#
#   3. If *not* in a Git repo but a `VERSION` file exists:
#         VERSION and BUILDNUMBER are read from that file (format: "<version> build <build_number>")
#
#   4. If neither a Git repo nor a VERSION file is present:
#         VERSION = <PACKAGE_VERSION> from configure
#         BUILDNUMBER = dev
#
#   5. If a BUILD_NUMBER file is present and BUILDNUMBER is "dev", it will override the default:
#         BUILDNUMBER = $(cat BUILD_NUMBER)
#
# Output:
#   - Default: <version> build <build_number>
#   - With --short: just <version>
#
# Examples:
#   ./getversion
#       → 2.0.1devel+dev.3.gabcd123 build dev
#
#   ./getversion --short
#       → 2.0.1devel+dev.3.gabcd123
#
#   (On exact tag)
#       → 2.0.0+RC1 build dev
#
# Supported Features:
#   - Handles annotated tags with special characters like `+` or `-`
#   - Produces semver-compatible version strings for CI/CD
#   - Works with or without Git
#
# Intended for use in:
#   - Release packaging
#   - Diagnostic reporting
#   - Embedding traceable version info into builds
#
# ======================================================================

# Make sure we're running from the root git repository
pushd "$(dirname "$0")" > /dev/null

# Extract PACKAGE_VERSION from the configure script
VERSION=$(perl -ne 'print $1 if /^PACKAGE_VERSION='\''(.+)'\''$/' < configure)
BUILDNUMBER=dev

# Function to generate a dev suffix like "dev.1.gabcd123"
generate_dev_version() {
    local latest_tag
    latest_tag=$(git describe --tags --abbrev=0)
    local full_desc
    full_desc=$(git describe --tags --long)
    local suffix="${full_desc#$latest_tag-}"
    echo "dev.${suffix//-/.}"
}

# Check if we're in a Git repo and git is available
if type git >/dev/null 2>&1 && [ -d '.git' ]; then
    # Ensure git describe doesn't fail due to shallow clone
    if git describe --tags --long >/dev/null 2>&1; then
        if git describe --exact-match >/dev/null 2>&1; then
            # We're exactly on a tag
            VERSION=$(git describe --exact-match)
        else
            # Not on a tag: add dev version suffix
            VERSION+="+$(generate_dev_version)"
        fi
    fi
# Not a git repo but VERSION file exists
elif [[ -s ./VERSION ]]; then
    VERSION=$(awk -F' build ' '{print $1}' ./VERSION)
    BUILDNUMBER=$(awk -F' build ' '{print $2}' ./VERSION)
fi

# Handle optional flag
FLAG="${1:-noflag}"
if [ "$FLAG" = "--short" ]; then
    echo "${VERSION}"
else
    if [[ ${BUILDNUMBER} = "dev" && -f BUILD_NUMBER ]]; then
        BUILDNUMBER=$(<BUILD_NUMBER)
    fi
    echo "${VERSION} build ${BUILDNUMBER}"
fi

popd > /dev/null
```

tuhaihe commented:
> @tuhaihe Here is my version of getversion. There was an issue if previous annotated tag had a "+" character.
> 
> example: git tag -a 2.0.0+RC1 -m "Apache Cloudberry (Incubating) 2.0.0 RC1"
> 
> This version of the script handles that case. Please review when you get a chance.

Hi @edespino thanks. I tested the different versions' logics, and your new file worked very well. So I replaced the old file with yours and added you as the co-author in the commit.

tuhaihe commented:
We can wait for PR #1081 for a few days to see the final result. Then will rebase the configure file.

xinzweb commented on configure:
```diff
@@ -1,6 +1,6 @@
 #! /bin/sh
 # Guess values for system-dependent variables and create Makefiles.
-# Generated by GNU Autoconf 2.69 for Apache Cloudberry 1.0.0.
+# Generated by GNU Autoconf 2.69 for Apache Cloudberry 2.0.0devel.
```
Are we using semver? If we using semver, then, this is not a correct version.

If we want to indicate a `-devel` release before the `2.0.0` official release, then we should have:

`2.0.0-devel` as the release version.

tuhaihe replied:
Just want to follow the PG way here. Hey, @edespino any ideas? Would like to have your input.

yjhjstz replied:
Of the two, **2.0.0-devel** is the more standard and commonly used version number, especially when following [Semantic Versioning](https://semver.org/). The typical format is:

```
major.minor.patch-pre-release
```

For example:  
- `2.0.0-alpha`
- `2.0.0-beta`
- `2.0.0-rc1`
- `2.0.0-devel`

The format **2.0.0devel** is uncommon and does not conform to the recommended semantic versioning style.

### Summary
- Recommended: `2.0.0-devel` (standard, tool-compatible, easy to recognize)
- Not recommended: `2.0.0devel`

If you want compatibility with mainstream release, packaging, and dependency management tools, you should use `2.0.0-devel` with the hyphen.

tuhaihe replied:
Pleased to see the reviewers' feedback. It's fine to adapt a more common approach for the development and release status. I'll provide updates later. Thanks all!

yjhjstz commented on getversion:
```diff
@@ -1,46 +1,119 @@
+#
+# Examples:
+#   ./getversion
+#       → 2.0.1devel+dev.3.gabcd123 build dev
```
modify related version to `2.0.1-devel`

tuhaihe replied:
Good catch! Although this is just an example, I will update it later for more accuracy.
